File: 1.1 The Elements of Programming/1-03.js
// Exercise 1.3
// Define a procedure that takes three numbers as arguments and returns the sum of the squares of the two larger numbers.

// Auxiliary function
var sumOfSquares = function (a, b) {
    return a*a + b*b;
};


// First solution
var sumOfSquaresOfBiggestNumbers = function (x, y, z) {
    var max = Math.max,
        min = Math.min;

    return sumOfSquares(max(x, y), max(min(x, y), z));
};

console.log(sumOfSquaresOfBiggestNumbers(1, 2, 3));


// Second solution
var sumOfSquaresOfBiggestNumbers2 = function (x, y, z) {
    var biggestNumbers = [].slice.call(arguments).sort();

    return sumOfSquares(biggestNumbers[1], biggestNumbers[2]);
};

console.log(sumOfSquaresOfBiggestNumbers2(1, 2, 3));


// Third solution
var sumOfSquaresOfBiggestNumbers3 = function (x, y, z) {
    return [].slice.call(arguments)
        .sort().slice(1)
        .reduce(function (a, b) {
            return a*a + b*b;
        });
};

console.log(sumOfSquaresOfBiggestNumbers3(1, 2, 3));

File: 1.1 The Elements of Programming/1-03.hs
-- Exercise 1.3
-- Define a procedure that takes three numbers as arguments and returns the sum of the squares of the two larger numbers.

import Data.List
import Test.HUnit

-- Auxiliary function
sumOfSquares :: (Num a) => a -> a -> a
sumOfSquares x y = x^2 + y^2


-- First solution, Java-like
f :: Int -> Int -> Int -> Int
f x y z
    | smallestNumber == x = sumOfSquares y z
    | smallestNumber == y = sumOfSquares x z
    | smallestNumber == z = sumOfSquares x y
    where smallestNumber = minimum [x, y, z]


-- Second solution, LISP-like
f' :: (Num a, Ord a) => a -> a -> a -> a
f' a b c = sumOfSquares x y
    where x = max a b
          y = max (min a b) c


-- Third solution, algebra-like
f'' :: (Num a, Ord a) => a -> a -> a -> a
f'' a b c = (sum . map square $ numbers) - (square $ minimum numbers)
    where square = (^2)
          numbers = [a, b, c]


-- Fourth solution, Haskell-like
f''' :: (Num a, Ord a) => a -> a -> a -> a
f''' a b c = sum . map (^2) . drop 1 . sort $ [a, b, c]


-- Tests
tests = TestList
    [ TestCase (assertEqual "first solution"  (f    1 2 3) 13)
    , TestCase (assertEqual "second solution" (f'   1 2 3) 13)
    , TestCase (assertEqual "third solution"  (f''  1 2 3) 13)
    , TestCase (assertEqual "fourth solution" (f''' 1 2 3) 13) ]

main = runTestTT tests